File: backend/app/api/traces.py
"""Admin trace exploration API for debugging workflows and tool executions."""
import os
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from pydantic import BaseModel, Field
import httpx

from app.db.session import get_db
from app.models.models import User, UserRole
from app.api.auth_deps import get_approved_user


router = APIRouter(prefix="/admin/traces", tags=["admin"])


# ============================================================================
# Response Models
# ============================================================================


class TracePreview(BaseModel):
    """Trace preview for list view."""
    trace_id: str
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    name: Optional[str] = None
    status: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    duration_ms: Optional[float] = None
    input_preview: Optional[str] = Field(None, max_length=200)
    output_preview: Optional[str] = Field(None, max_length=200)
    tags: Optional[List[str]] = []


class TraceListResponse(BaseModel):
    """Paginated trace list response."""
    traces: List[TracePreview]
    total: int
    limit: int
    offset: int
    has_more: bool = False


class SpanDetail(BaseModel):
    """Detailed span information."""
    span_id: str
    trace_id: str
    parent_span_id: Optional[str] = None
    name: str
    start_time: datetime
    end_time: Optional[datetime] = None
    duration_ms: Optional[float] = None
    status: Optional[str] = None
    input: Optional[Dict[str, Any]] = None
    output: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None
    level: Optional[str] = None


class TraceDetail(BaseModel):
    """Full trace details."""
    trace_id: str
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    name: Optional[str] = None
    status: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    duration_ms: Optional[float] = None
    input: Optional[Dict[str, Any]] = None
    output: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None
    tags: Optional[List[str]] = []
    observations: Optional[List[SpanDetail]] = []


class TraceSearchRequest(BaseModel):
    """Search request with filters."""
    user_id: Optional[str] = None
    session_id: Optional[str] = None
    name: Optional[str] = None
    status: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    tags: Optional[List[str]] = None
    limit: int = Field(default=50, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


# ============================================================================
# Helper Functions
# ============================================================================


def verify_admin(current_user: User):
    """Verify user has admin role."""
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=403,
            detail="Admin access required"
        )


def get_langfuse_client() -> httpx.Client:
    """Get configured Langfuse HTTP client (synchronous)."""
    host = os.getenv("LANGFUSE_HOST", "http://localhost:3001")
    public_key = os.getenv("LANGFUSE_PUBLIC_KEY", "")
    secret_key = os.getenv("LANGFUSE_SECRET_KEY", "")

    if not public_key or not secret_key:
        raise HTTPException(
            status_code=500,
            detail="Langfuse credentials not configured"
        )

    return httpx.Client(
        base_url=host,
        auth=(public_key, secret_key),
        timeout=10.0
    )


def sanitize_preview(data: Any, max_length: int = 200) -> Optional[str]:
    """Sanitize and truncate data for preview."""
    if data is None:
        return None

    # Convert to string
    text = str(data)

    # Remove PII patterns (basic email/phone scrubbing)
    # TODO: Add more sophisticated PII detection
    import re
    text = re.sub(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', '[EMAIL]', text)
    text = re.sub(r'\b\d{3}[-.]?\d{3}[-.]?\d{4}\b', '[PHONE]', text)

    # Truncate
    if len(text) > max_length:
        return text[:max_length] + "..."

    return text


def filter_by_tenant(traces: List[Dict], user: User, db: Session) -> List[Dict]:
    """Filter traces to only include those from user's tenant."""
    if not user.tenant_id:
        return traces

    # Get all user IDs in the tenant
    from app.models.models import User as UserModel
    tenant_user_ids = [
        str(u.id) for u in db.query(UserModel)
        .filter(UserModel.tenant_id == user.tenant_id)
        .all()
    ]

    # Filter traces
    filtered = []
    for trace in traces:
        trace_user_id = trace.get("userId") or trace.get("user_id")
        if trace_user_id and trace_user_id in tenant_user_ids:
            filtered.append(trace)

    return filtered


# ============================================================================
# API Endpoints
# ============================================================================


@router.get("", response_model=TraceListResponse)
async def list_traces(
    trace_id: Optional[str] = Query(None, description="Filter by trace ID"),
    user_id: Optional[str] = Query(None, description="Filter by user ID"),
    session_id: Optional[str] = Query(None, description="Filter by session ID"),
    name: Optional[str] = Query(None, description="Filter by trace name"),
    status: Optional[str] = Query(None, description="Filter by status"),
    start_date: Optional[datetime] = Query(None, description="Start date for date range"),
    end_date: Optional[datetime] = Query(None, description="End date for date range"),
    limit: int = Query(50, ge=1, le=100, description="Max traces to return"),
    offset: int = Query(0, ge=0, description="Pagination offset"),
    current_user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    """
    List traces with filtering and pagination.

    Requires: Admin role.

    Returns paginated list of traces with preview data only.
    Full trace details available via GET /traces/{trace_id}.
    """
    verify_admin(current_user)

    # Build query parameters
    params: Dict[str, Any] = {
        "limit": limit,
        "offset": offset,
    }

    if trace_id:
        params["traceId"] = trace_id
    if user_id:
        params["userId"] = user_id
    if session_id:
        params["sessionId"] = session_id
    if name:
        params["name"] = name
    if status:
        params["status"] = status
    if start_date:
        params["fromTimestamp"] = start_date.isoformat()
    if end_date:
        params["toTimestamp"] = end_date.isoformat()

    # Query Langfuse API
    with get_langfuse_client() as client:
        try:
            response = client.get("/api/public/traces", params=params)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPError as e:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to query Langfuse: {str(e)}"
            )

    # Extract traces and track original total before filtering
    traces_data = data.get("data", [])
    total_before_filter = data.get("meta", {}).get("totalItems", len(traces_data))

    # Filter by tenant
    traces_data = filter_by_tenant(traces_data, current_user, db)

    # Convert to preview format
    traces = []
    for trace in traces_data:
        traces.append(TracePreview(
            trace_id=trace.get("id", ""),
            user_id=trace.get("userId"),
            session_id=trace.get("sessionId"),
            name=trace.get("name"),
            status=trace.get("status"),
            created_at=datetime.fromisoformat(trace["timestamp"].replace("Z", "+00:00")),
            updated_at=datetime.fromisoformat(trace["updatedAt"].replace("Z", "+00:00")) if trace.get("updatedAt") else None,
            duration_ms=trace.get("duration"),
            input_preview=sanitize_preview(trace.get("input")),
            output_preview=sanitize_preview(trace.get("output")),
            tags=trace.get("tags", []),
        ))

    # Calculate has_more based on whether we could fetch more if we tried
    has_more = (offset + len(traces)) < total_before_filter

    return TraceListResponse(
        traces=traces,
        total=len(traces),  # Number of traces returned in this batch
        limit=limit,
        offset=offset,
        has_more=has_more,
    )


@router.get("/{trace_id}", response_model=TraceDetail)
async def get_trace(
    trace_id: str,
    current_user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    """
    Get full details for a single trace.

    Requires: Admin role.

    Returns complete trace data including all observations/spans.
    """
    verify_admin(current_user)

    # Query Langfuse API
    with get_langfuse_client() as client:
        try:
            response = client.get(f"/api/public/traces/{trace_id}")
            response.raise_for_status()
            trace = response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                raise HTTPException(status_code=404, detail="Trace not found")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to fetch trace: {str(e)}"
            )
        except httpx.HTTPError as e:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to query Langfuse: {str(e)}"
            )

    # Verify tenant access
    filtered = filter_by_tenant([trace], current_user, db)
    if not filtered:
        raise HTTPException(
            status_code=403,
            detail="Trace not found or access denied"
        )

    # Convert observations to spans
    observations = trace.get("observations", [])
    spans = []
    for obs in observations:
        spans.append(SpanDetail(
            span_id=obs.get("id", ""),
            trace_id=trace_id,
            parent_span_id=obs.get("parentObservationId"),
            name=obs.get("name", ""),
            start_time=datetime.fromisoformat(obs["startTime"].replace("Z", "+00:00")),
            end_time=datetime.fromisoformat(obs["endTime"].replace("Z", "+00:00")) if obs.get("endTime") else None,
            duration_ms=obs.get("duration"),
            status=obs.get("level"),
            input=obs.get("input"),
            output=obs.get("output"),
            metadata=obs.get("metadata"),
            level=obs.get("level"),
        ))

    return TraceDetail(
        trace_id=trace.get("id", ""),
        user_id=trace.get("userId"),
        session_id=trace.get("sessionId"),
        name=trace.get("name"),
        status=trace.get("status"),
        created_at=datetime.fromisoformat(trace["timestamp"].replace("Z", "+00:00")),
        updated_at=datetime.fromisoformat(trace["updatedAt"].replace("Z", "+00:00")) if trace.get("updatedAt") else None,
        duration_ms=trace.get("duration"),
        input=trace.get("input"),
        output=trace.get("output"),
        metadata=trace.get("metadata"),
        tags=trace.get("tags", []),
        observations=spans,
    )


@router.get("/{trace_id}/spans", response_model=List[SpanDetail])
async def get_trace_spans(
    trace_id: str,
    current_user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    """
    Get all spans/observations within a trace.

    Requires: Admin role.

    Returns array of spans with timing, status, and I/O data.
    """
    verify_admin(current_user)

    # Get full trace (includes observations)
    trace_detail = await get_trace(trace_id, current_user, db)

    return trace_detail.observations or []


@router.post("/search", response_model=TraceListResponse)
async def search_traces(
    search_request: TraceSearchRequest,
    current_user: User = Depends(get_approved_user),
    db: Session = Depends(get_db),
):
    """
    Advanced trace search with filters.

    Requires: Admin role.

    Accepts JSON body with multiple filter criteria.
    Returns matching traces with pagination.
    """
    verify_admin(current_user)

    # Build query parameters from request body
    params: Dict[str, Any] = {
        "limit": search_request.limit,
        "offset": search_request.offset,
    }

    if search_request.user_id:
        params["userId"] = search_request.user_id
    if search_request.session_id:
        params["sessionId"] = search_request.session_id
    if search_request.name:
        params["name"] = search_request.name
    if search_request.status:
        params["status"] = search_request.status
    if search_request.start_date:
        params["fromTimestamp"] = search_request.start_date.isoformat()
    if search_request.end_date:
        params["toTimestamp"] = search_request.end_date.isoformat()
    if search_request.tags:
        # Tags filter (Langfuse API may support this differently)
        params["tags"] = ",".join(search_request.tags)

    # Query Langfuse API
    with get_langfuse_client() as client:
        try:
            response = client.get("/api/public/traces", params=params)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPError as e:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to query Langfuse: {str(e)}"
            )

    # Extract traces and track original total
    traces_data = data.get("data", [])
    total_before_filter = data.get("meta", {}).get("totalItems", len(traces_data))

    # Filter by tenant
    traces_data = filter_by_tenant(traces_data, current_user, db)

    # Apply tag filtering if needed (client-side)
    if search_request.tags:
        traces_data = [
            t for t in traces_data
            if any(tag in t.get("tags", []) for tag in search_request.tags)
        ]

    # Convert to preview format
    traces = []
    for trace in traces_data:
        traces.append(TracePreview(
            trace_id=trace.get("id", ""),
            user_id=trace.get("userId"),
            session_id=trace.get("sessionId"),
            name=trace.get("name"),
            status=trace.get("status"),
            created_at=datetime.fromisoformat(trace["timestamp"].replace("Z", "+00:00")),
            updated_at=datetime.fromisoformat(trace["updatedAt"].replace("Z", "+00:00")) if trace.get("updatedAt") else None,
            duration_ms=trace.get("duration"),
            input_preview=sanitize_preview(trace.get("input")),
            output_preview=sanitize_preview(trace.get("output")),
            tags=trace.get("tags", []),
        ))

    # Calculate has_more
    has_more = (search_request.offset + len(traces)) < total_before_filter

    return TraceListResponse(
        traces=traces,
        total=len(traces),
        limit=search_request.limit,
        offset=search_request.offset,
        has_more=has_more,
    )

File: frontend/lib/traces.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// ==================== Type Definitions ====================

export interface TraceFilter {
  trace_id?: string;
  user_id?: string;
  session_id?: string;
  name?: string;
  status?: 'success' | 'error' | 'timeout' | 'validation_error';
  start_date?: string;  // ISO datetime
  end_date?: string;
  limit?: number;
  offset?: number;
}

export interface TracePreview {
  trace_id: string;
  user_id: string | null;
  session_id: string | null;
  name: string | null;
  status: string;
  created_at: string;
  duration_ms: number | null;
  input_preview: string | null;
  output_preview: string | null;
}

export interface TraceSpan {
  id: string;
  type: string;
  name: string | null;
  start_time: string;
  end_time: string | null;
  input: Record<string, any> | null;
  output: Record<string, any> | null;
  metadata: Record<string, any> | null;
}

export interface TraceDetail {
  trace_id: string;
  user_id: string | null;
  session_id: string | null;
  name: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  duration_ms: number | null;
  input: Record<string, any> | null;
  output: Record<string, any> | null;
  metadata: Record<string, any> | null;
  observations: TraceSpan[];
}

export interface TraceListResponse {
  traces: TracePreview[];
  total: number;
  limit: number;
  offset: number;
}

export interface TestResultFilter {
  scenario_name?: string;
  status?: 'passed' | 'failed' | 'error';
  limit?: number;
  offset?: number;
}

export interface TestResult {
  id: number;
  test_run_id: number;
  scenario_name: string;
  status: string;
  duration_seconds: number | null;
  error_message: string | null;
  trace_id: string | null;
  created_at: string;
}

export interface TestResultsResponse {
  results: TestResult[];
  total: number;
  limit: number;
  offset: number;
}

// ==================== Traces API Client ====================

class TracesApiClient {
  private baseURL: string;
  private token: string | null = null;

  constructor() {
    this.baseURL = API_URL;

    // Load token from localStorage if available
    if (typeof window !== 'undefined') {
      this.token = localStorage.getItem('access_token');
    }
  }

  setToken(token: string) {
    this.token = token;
  }

  /**
   * Internal request helper method
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    // Add token to request headers
    if (this.token) {
      (headers as Record<string, string>)['Authorization'] = `Bearer ${this.token}`;
    } else if (typeof window !== 'undefined') {
      // Try to load from localStorage if not in memory
      const storedToken = localStorage.getItem('access_token');
      if (storedToken) {
        (headers as Record<string, string>)['Authorization'] = `Bearer ${storedToken}`;
      }
    }

    const response = await fetch(`${this.baseURL}${endpoint}`, {
      ...options,
      headers,
    });

    if (!response.ok) {
      let errorMessage = `HTTP error! status: ${response.status}`;
      try {
        const error = await response.json();
        errorMessage = error.detail || error.message || errorMessage;
      } catch {
        // If response is not JSON, use default message
      }

      // Handle specific error cases
      if (response.status === 401) {
        throw new Error('Unauthorized: Admin access required');
      }
      if (response.status === 404) {
        throw new Error(`Not found: ${errorMessage}`);
      }

      throw new Error(errorMessage);
    }

    return response.json();
  }

  /**
   * List traces with optional filters and pagination
   */
  async getTraces(filters: TraceFilter = {}): Promise<TraceListResponse> {
    const params = new URLSearchParams();

    if (filters.trace_id) params.append('trace_id', filters.trace_id);
    if (filters.user_id) params.append('user_id', filters.user_id);
    if (filters.session_id) params.append('session_id', filters.session_id);
    if (filters.name) params.append('name', filters.name);
    if (filters.status) params.append('status', filters.status);
    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters.offset !== undefined) params.append('offset', filters.offset.toString());

    const queryString = params.toString();
    const endpoint = `/api/admin/traces${queryString ? `?${queryString}` : ''}`;

    try {
      return await this.request<TraceListResponse>(endpoint);
    } catch (error) {
      throw new Error(`Failed to fetch traces: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Get a single trace with full details and observations/spans
   */
  async getTrace(traceId: string): Promise<TraceDetail> {
    if (!traceId) {
      throw new Error('trace_id is required');
    }

    try {
      return await this.request<TraceDetail>(`/api/admin/traces/${traceId}`);
    } catch (error) {
      throw new Error(`Failed to fetch trace ${traceId}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Search traces by correlation_id or other query parameters
   */
  async searchTraces(query: string, filters: Omit<TraceFilter, 'trace_id'> = {}): Promise<TraceListResponse> {
    if (!query) {
      throw new Error('query parameter is required');
    }

    const params = new URLSearchParams();
    params.append('q', query);

    if (filters.user_id) params.append('user_id', filters.user_id);
    if (filters.session_id) params.append('session_id', filters.session_id);
    if (filters.name) params.append('name', filters.name);
    if (filters.status) params.append('status', filters.status);
    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters.offset !== undefined) params.append('offset', filters.offset.toString());

    const queryString = params.toString();
    const endpoint = `/api/admin/traces/search?${queryString}`;

    try {
      return await this.request<TraceListResponse>(endpoint);
    } catch (error) {
      throw new Error(`Failed to search traces: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Get test results with optional filters and pagination
   */
  async getTestResults(filters: TestResultFilter = {}): Promise<TestResultsResponse> {
    const params = new URLSearchParams();

    if (filters.scenario_name) params.append('scenario_name', filters.scenario_name);
    if (filters.status) params.append('status', filters.status);
    if (filters.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters.offset !== undefined) params.append('offset', filters.offset.toString());

    const queryString = params.toString();
    const endpoint = `/api/admin/test-results${queryString ? `?${queryString}` : ''}`;

    try {
      return await this.request<TestResultsResponse>(endpoint);
    } catch (error) {
      throw new Error(`Failed to fetch test results: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Get a single test result
   */
  async getTestResult(testId: string | number): Promise<TestResult> {
    if (!testId) {
      throw new Error('test_id is required');
    }

    try {
      return await this.request<TestResult>(`/api/admin/test-results/${testId}`);
    } catch (error) {
      throw new Error(`Failed to fetch test result ${testId}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
}

// ==================== Export ====================

export const tracesApi = new TracesApiClient();
